Waciuma/tutorials (#772)
vmware-tanzu/cartographer: site/content/docs/development/tutorials/runnable.md

here](https://github.com/pivotal/kpack/blob/main/docs/install.md).

Alternatively, you may choose to use the
[./hack/setup.sh](https://github.com/vmware-tanzu/cartographer/blob/main/hack/setup.sh)
script to install a kind cluster with Cartographer and Tekton. _This script is 
meant for our end-to-end testing and while we rely
on it working in that role, no user guarantees are made about the script._
Command to run from the Cartographer directory:
```shell
$ ./hack/setup.sh cluster cartographer-latest example-dependencies
```

If you later wish to tear down this generated cluster, run
```shell
$ ./hack/setup.sh teardown
```

## Scenario
Our CTO is interested in putting quality controls in place; only code that
passes certain checks should be built and deployed. They want to start small:
all source code repositories that are built must pass markdown linting. In
order to do this we’re going to leverage [the markdown linting pipeline in
the TektonCD
catalog](https://github.com/tektoncd/catalog/tree/main/task/markdown-lint/0.1).

In this tutorial we’ll see how to use Cartographer’s Runnable to give
us easy updating behavior of Tekton (no need for Tekton Triggers and Github

> Webhooks). In [the following tutorial](runnable-in-a-supply-chain.md) we’ll

Reviewer: ```suggestion
Webhooks). In [the next tutorial](runnable-in-a-supply-chain.md) we’ll
```

I wonder if this is a regional thing, but 'following' here, because we're just starting this tutorial, means "what we're about to do" to me.

```

Finally we need to create a pipeline-run object. This object provides the
param and workspace values defined at the top level `.spec` field of the
pipeline.

```yaml
---
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: linter-pipeline-run
spec:
  pipelineRef:
    name: linter-pipeline
  params:
    - name: repository
      value: https://github.com/waciumawanjohi/demo-hello-world
    - name: revision
      value: main
  workspaces:
    - name: shared-workspace
      volumeClaimTemplate:
        spec:
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 256Mi
```

Importantly, this pipeline-run object will kick off a single run
of the pipeline. The run will either succeed or fail. The outcome will be
written into the pipeline-run’s status. No later changes to the pipeline-run
object will change those outcomes; the run happens once.

To see this in action, let’s apply the above pipeline-run. If we watch the
object, we’ll soon see that it succeeds.

```shell
$ watch 'kubectl get -o yaml pipelinerun linter-pipeline-run | yq .status.conditions'
```

Eventually yields the result:

> ```json
> [
>   {
>     "lastTransitionTime": ...,
>     "message": "Tasks Completed: 2 (Failed: 0, Cancelled 0), Skipped: 0",
>     "reason": "Succeeded",
>     "status": "True",
>     "type": "Succeeded"
>   }
> ]
> ```

Reviewer 2: this command is using the non-standard version of `yq`. the more standard version (which more people will likely be using) will produce this output:
```suggestion
```yaml
- lastTransitionTime: ...
  message: "Tasks Completed: 2 (Failed: 0, Cancelled 0), Skipped: 0"
  reason: "Succeeded"
  status: "True"
  type: "Succeeded"
```
```

documentation](https://tekton.dev/docs/).

Now we examine the Cartographer Runnable object. We expect it to expose values
from our successful object.

```shell
$ kubectl get -o yaml runnable linter
```

```yaml
apiVersion: carto.run/v1alpha1
kind: Runnable
Metadata: ...
Spec: ...
status:
  conditions: ...
  observedGeneration: ...
  outputs:
    lastTransitionTime: "2022-03-07T19:24:35Z"
```

Wonderful! The value from the field we specified in the ClusterRunTemplate
is now in the outputs of the Runnable.

Finally, let's update our runnable with a new repository:

```yaml
apiVersion: carto.run/v1alpha1
kind: Runnable
metadata:
  name: linter
spec:
  runTemplateRef:
    name: md-linting-pipelinerun
  inputs:
    repository: https://github.com/kelseyhightower/nocode # <=== new repo
    revision: master # <=== new revision
  serviceAccountName: pipeline-run-management-sa
```

When we apply this to the cluster, we can observe:

- The spec of our runnable is updated
- The runnable causes the creation of a new pipelinerun
- The new pipeline run fails because the new repo does not pass linting
- Since the result of the new pipeline run is failure, our runnable output

>   remains the output of our previous (successful pipeline run)

Reviewer 2: ```suggestion
  remains the output of our previous (successful) pipeline run
```

> ## Wrapup

Reviewer 2: nit
```suggestion
## Wrap Up
```

In this tutorial you learned:
- Some useful kubernetes objects cannot be updated

> - Runnable allows you to add update-able behavior to those objects

Reviewer 2: to be consistent with how it's spelled elsewhere in this and the next tutorial
```suggestion
- Runnable allows you to add updateable behavior to those objects
```
